9511 Export Enums as "Type"

incendiary-ca opened the pull request:
Change typedef codegen to export the underlying enum itself, rather
than an alias. Works around https://github.com/rust-lang/rust/issues/31355

Fixes #9511.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9691)

<!-- Reviewable:end -->

incendiary-ca committed:
9511 Export Enums as "Type"

Change typedef codegen to export the underlying enum itself, rather
than an alias. Works around https://github.com/rust-lang/rust/issues/31355

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

KiChjang commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -5324,14 +5324,15 @@ def __init__(self, config, prefix, webIDLFile):
+                                    (t.identifier.name)))])
+                else:
+                    assert not typeNeedsRooting(t.innerType, config.getDescriptorProvider)
+                    cgthings.extend([CGGeneric("\npub type %s = " % (t.identifier.name)),
```
Even if it's not a Union typedef, I think we should still re-export it, instead of using the `pub type = blah` syntax.

incendiary-ca committed:
Export Enums

Fix codegen to export non-Union typedefs using
pub use self::enum as type;

KiChjang commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -5324,14 +5324,14 @@ def __init__(self, config, prefix, webIDLFile):
-                cgthings.extend([CGGeneric("\npub type %s = " % (t.identifier.name)),
-                                 getRetvalDeclarationForType(t.innerType, config.getDescriptorProvider()),
-                                 CGGeneric(";\n\n")])
+            for e in enums:
```
I don't think this is limited to just enums. We want all typedefs to be re-exported.

KiChjang commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -5324,14 +5324,14 @@ def __init__(self, config, prefix, webIDLFile):
-                                 CGGeneric(";\n\n")])
+            for e in enums:
+                if t.innerType.isUnion():
+                    cgthings.extend([CGGeneric("\npub use self::%s as %s;\n\n" % (e.identifier.name,
```
Since this is a `use` statement, it needs to be put at the top of the file, and in alphabetical order with other `use` statements as well.

KiChjang commented:
I don't think I'm qualified enough to review Python code. r? @nox

nox commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 5329 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9691#-KApkJKNHqQ-VuyzgW6v-r2-5329:-KBIHT17N3zIWEnhSNXl:-2041452902) ([raw file](https://github.com/servo/servo/blob/d8bf193381bc26127607b2d22c6eceee6b278bfe/components/script/dom/bindings/codegen/CodegenRust.py#L5329)):*
In codegen, this does not matter.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9691)*
<!-- Sent from Reviewable.io -->

Tycheon commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 5329 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9691#-KApkJKNHqQ-VuyzgW6v-r2-5329:-KBKtO0rbpiDcCFNvFji:-2001523477) ([raw file](https://github.com/servo/servo/blob/d8bf193381bc26127607b2d22c6eceee6b278bfe/components/script/dom/bindings/codegen/CodegenRust.py#L5329)):*
If not, then is this good to go, or is there something left outstanding?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9691)*
<!-- Sent from Reviewable.io -->

KiChjang commented:
@Tycheon One comment still needs to be addressed.

Tycheon commented:
Photography has gotten quite busy this past week, so I've not really had any time to look at this. I'll do some more digging tomorrow to try and address that one remaining comment (that it should hit *all* typedefs and not just typedef'd enums, correct?)

That being said, I didn't see anything in the CodeGenRust.py that was iterating over typedefs that were NOT enums ... so, if I didn't miss it, either there aren't any typedefs that aren't for enums, or they weren't being exported, or they were being exported somewhere else.

Tycheon commented:
pub use Option<i32> as Bar;

This will give an error. Using the existing method of:

pub type Bar = Option<i32>;

Does work.

Given that some of the typedefs are of this form (not sure if it's just the nonUnion typedefs just yet), I won't be able to change the export for **all** typedefs.

I'll figure out if this affects the Union typedefs (I've a feeling it doesn't), and fix an error I made in my initial PR before submitting another one.

nox commented:
Only union typedefs should use reexporting, yes

Tycheon commented:
... was definitely overthinking things. Should be a new PR in the queue for changes to lines 5352 and 5353 of CodegenRust.py which exports the typedefs using the UnionTypes provided in dom::bindings::codegen, rather than just incorrectly exporting the local enum.

Commit:
Merge remote-tracking branch 'origin' into 9511

Commit:
9511 Export Typedefs Using "pub use" Syntax

Fix codegen to export typedefs using "pub use foo as bar;" instead of
"pub type bar = foo;", as the latter breaks with enums.

Commit:
Fix Travis CI line length error

nox commented:
Please rebase and squash. AS for the tidy line length fix, just return to a new line after `%`.

jdm commented:
Rebased in #10152.
